src/vs/base/node/powershell.ts

```diff
@@ -0,0 +1,326 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as pfs from 'vs/base/node/pfs';
+import * as os from 'os';
+import * as path from 'vs/base/common/path';
+import { env } from 'vs/base/common/process';
+
+const WindowsPowerShell64BitLabel = 'Windows PowerShell';
+const WindowsPowerShell32BitLabel = 'Windows PowerShell (x86)';
+
+// This is required, since parseInt("7-preview") will return 7.
+const IntRegex: RegExp = /^\d+$/;
+
+const PwshMsixRegex: RegExp = /^Microsoft.PowerShell_.*/;
+const PwshPreviewMsixRegex: RegExp = /^Microsoft.PowerShellPreview_.*/;
+
+// The platform details descriptor for the platform we're on
+const isProcess64Bit: boolean = process.arch === 'x64';
+const isOS64Bit: boolean = isProcess64Bit || env.hasOwnProperty('PROCESSOR_ARCHITEW6432');
+
+export interface IPowerShellExeDetails {
+	readonly displayName: string;
+	readonly exePath: string;
+}
+
+export interface IPossiblePowerShellExe extends IPowerShellExeDetails {
+	exists(): Promise<boolean>;
+}
+
+class PossiblePowerShellExe implements IPossiblePowerShellExe {
+	public readonly exePath: string;
+	public readonly displayName: string;
+
+	private knownToExist: boolean | undefined;
+
+	constructor(
+		pathToExe: string,
+		installationName: string,
+		{ knownToExist = false }: { knownToExist?: boolean } = {}) {
+
+		this.exePath = pathToExe;
+		this.displayName = installationName;
+		this.knownToExist = knownToExist || undefined;
+	}
+
+	public async exists(): Promise<boolean> {
+		if (this.knownToExist === undefined) {
+			this.knownToExist = await pfs.fileExists(this.exePath);
+		}
+		return this.knownToExist;
+	}
+}
+
+function getProgramFilesPath(
+	{ useAlternateBitness = false }: { useAlternateBitness?: boolean } = {}): string | null {
+
+	if (!useAlternateBitness) {
+		// Just use the native system bitness
+		return env.ProgramFiles!;
```

Comment:
Is it worth checking these or are they guaranteed to exist? We could avoid doing the `!` with:

```ts
return env.ProgramFiles || null;
```

Reply:
missed this comment. Good idea. Changed.